File: threeSum.py
# -*- coding:utf-8 -*- 

#https://leetcode.com/problems/3sum/description/
class Solution(object):
	
	# for
	def threeSum(self,nums):
		result =[]
		if len(nums) == 0:
			return result

		#sor	
		nums.sort()
		leg=len(nums)
		for i in range(0,leg):
			# 相同的key，同一key不要再循环了，防止重复
			if i>0 and nums[i]==nums[i-1]:
				continue
			for j in range(i+1,leg):
				if j>i+1 and nums[j]==nums[j-1]:
					continue
				for k in range(j+1,leg):
					if k>j+1 and nums[k]==nums[k-1]:
						continue
					if (nums[i]+nums[j]+nums[k])==0:
						group=[nums[i],nums[j],nums[k]]
						result.append(group)	
		return result

	#a+b=-c		
	def threeSum2(self,nums):
		result =[]
		leg =len(nums)
		nums.sort()
		for i in range(0,leg-2):
			if i==0 or (i>0 and nums[i]!=nums[i-1]):
				left=i+1
				right=leg-1
				summ =0-nums[i]
				# print(summ)
				while left<right:
					if (nums[left]+nums[right])==summ:
						group=[nums[left],nums[i],nums[right]]
						result.append(group)
						while (left<right and nums[left]==nums[left+1]):
							left=left+1
						while (left<right and nums[right] ==nums[right-1]):
							right=right-1
						left=left+1
						right=right-1
					elif (nums[left]+nums[right])<summ:
						left=left+1
					else:
						right=right-1
		return result

if __name__ == "__main__":
	print("hell world")
	# S = [-1, 0, 1, 1,1,1,2, -1,-1,-1, -4]
	S = [-1, 0, 1, 2, -1, -4]
	solution = Solution();
	result=solution.threeSum2(S);
	print(result)

File: leetcode_java/leecode/ThreeSum.java
package leecode;

import com.sun.deploy.util.ArrayUtil;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * three sum 算法
 *
 * 给定一个n个整数的数组S，是否存在S中的元素a，b，c，使得a + b + c = 0？查找数组中所有唯一的三元组，它们的总和为零。
 * @author <a href="mailto:jian.wu@happyelements.com">jian.wu</a>
 * @version 1.0
 * @link https://leetcode.com/problems/3sum/description/
 * @since 1.0
 */
public class ThreeSum {

    public List<List<Integer>> threeSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (null == nums || nums.length == 0) {
            return res;
        }

        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            //剔除相同的，相同必然组合相同
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int left = i + 1;
            int right = nums.length - 1;
            int sum = target - nums[i];
            while (left < right) {
                int kSum = nums[left] + nums[right];
                if (kSum == sum) {
                    List<Integer> group = Arrays.asList(nums[i], nums[left], nums[right]);
                    res.add(group);

                    //剔除相同的数字，相同必然组合相同
                    while (left < right && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right - 1]) {
                        right--;
                    }

                    left++;
                    right--;
                } else if (kSum < sum) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return res;
    }

    public static void main(String[] args) {
//        int[] nums = {-1, 0, 1, 1, 1, 1, 2, -1, -1, -1, -4};
        int[] nums = {-1, -1, 0, 0, 1, 2, 2, 2, 2, -2, -4};

        ThreeSum threeSum = new ThreeSum();
        List<List<Integer>> res = threeSum.threeSum(nums, 0);
        for (List<Integer> sigRes : res) {
            for (int i : sigRes) {
                System.out.print(i);
                System.out.print(",");
            }
            System.out.println();
        }
    }
}
